### DISCUSSION TITLE
How to remove ramdom black pixels in picture?

### DISCUSSION THREAD

**Discussion by andreasm123** [ImageMagick/ImageMagick#4318] [Digital Image Processing] [answered]
I'm using convert on Linux (as part of ImageMagick) - does anyone know how to remove random black pixels in a picture? (it's a satellite image received with some random disturbance caused by rf-link noise)

**Comment by fmw42** [comment]
Please post an example image. Also what is your ImageMagick exact version?

**Comment by snibgo** [comment]
We need more information.

What defines which black pixels should be removed, as opposed to black pixels that should be left untouched?

What does "remove" a black pixel mean? Perhaps you want to make them transparent, or turn them red, or replace them with a local average.

**Comment by andreasm123** [comment]
Thanks for your replies ...

>What does "remove" a black pixel mean? Perhaps you want to make them transparent, or turn them red, or replace them with a local average.

My idea was to have them replaced with local average.

>Please post an example image. Also what is your ImageMagick exact version?

I've attached 2 example images below which illustrate that. My ImageMagick version is:

`Version: ImageMagick 6.9.7-4 Q16 arm 20170114 http://www.imagemagick.org
Copyright: © 1999-2017 ImageMagick Studio LLC
License: http://www.imagemagick.org/script/license.php
Features: Cipher DPC Modules OpenMP
Delegates (built-in): bzlib djvu fftw fontconfig freetype jbig jng jp2 jpeg lcms lqr ltdl lzma openexr pangocairo png tiff wmf x xml zlib
`

<img width="455" alt="2021-09-30_19_55_NOAA19" src="https://user-images.githubusercontent.com/82610383/136096500-c8d807c9-66da-4ab1-9472-a15e74f92e1e.png">
<img width="455" alt="2021-10-01_12_14_NOAA18" src="https://user-images.githubusercontent.com/82610383/136096510-24191cdb-1f38-4bd4-98a0-494054fb9956.png">

**Comment by fmw42** [comment, marked as answer]
Here is one way. It first does a median filter over the whole image. Then it creates a mask by thresholding on the darkest pixels in the original. Then it uses the mask to composite the two images together replacing only where the mask is white with the median values into the original image. You can adjust the threshold % and the size of the median filter.

```
convert 136096500-c8d807c9-66da-4ab1-9472-a15e74f92e1e.png \
\( -clone 0 -statistic median 13x13 \) \
\( -clone 0 -auto-level -threshold 1% -negate +write x.png \) \
-compose over -composite result1.png

``` 

<img width="455" alt="result1" src="https://user-images.githubusercontent.com/13458218/136099430-9a4dd891-eb86-41fa-abe9-bcba156d4e48.png">

I do not know how to recognize your dropped pixels from just very dark spots. So my result is rather conservative. If you know you only have perfectly black (graylevel 0) pixels for the dropped one, then select a 0%. The median filter should be chosen for the size of the dropped regions if they are larger than isolated single pixels. Otherwise, just use 3x3.

You could replace the median with an average, but the results would likely be more blurry if you have to replace large clumps.

**Reply by andreasm123** [reply]
Thanks - that helped !
